### ISSUE TITLE
how to use --enable-shared ?

### ISSUE THREAD

**Issue by buncis** [heroku/heroku-buildpack-python#1225]
I've tried to fork this repository and add `--enable-shared` to this line

https://github.com/heroku/heroku-buildpack-python/blob/main/builds/runtimes/python3#L26

```
./configure --prefix=$OUT_PREFIX --with-ensurepip=no --enable-shared
make
make install
```

but it not works my heroku builded without using the python that have enable-shared 
even I think heroku build the python without using my buildpacks at all 
any guidance how to use python3 that compiled using --enable-shared ?

**Comment by edmorley** [comment]
@buncis Hi! The scripts under `builds/` in this repository are not run when each app is built using the buildpack, but separately when releasing new Python versions to S3. You will need to run the compile scripts uploading to your own S3 bucket ([build docs]()), then use `BUILDPACK_S3_BASE_URL` to point at that bucket (after mirroring the other contents eg pip wheels). `BUILDPACK_S3_BASE_URL` cannot be set as an app env var, and will need to be set via an earlier buildpack's export script (or just fork the buildpack).

**Comment by buncis** [comment]
@edmorley thanks for the quick response

I see, so the correct way to enable-shared is just like this fork right ?

https://github.com/ReforgeHQ/heroku-buildpack-python/commit/813825b1d886aa78f70070cedcae3aeb1ea8fe46

I've tried that way too, I fork that fork and change the line to my desired python version but somehow its not working and I forced to use heroku-18 stack for that fork

ok please bear with me

so what I need to do is

[ ] compile the python myself with --enable-shared configuration
[ ] upload the compile result to amazon s3
[ ] fork this heroku-buildpack-python and change the `BUILDPACK_S3_BASE_URL` to my own s3

**Comment by edmorley** [comment]
> so what I need to do is

Yeah that's one approach.

Another would be to just compile during the build, and store the result in the cache, so you don't need to mess about with S3.

Example at:
https://github.com/heroku/heroku-buildpack-python/issues/488#issuecomment-659579578

(Though note this is all unsupported and you'll unfortunately mostly need to figure this out yourself)

**Comment by buncis** [comment]
I think I will follow that #488 approach since its almost have same use case and simpler

but I don't know where to put the `post_compile` scripts

did this is the correct file ?
https://github.com/heroku/heroku-buildpack-python/blob/f3abce6ac660a009c7ffac95d5b86d231555c28a/bin/steps/hooks/post_compile

yeah If I could pass this 
I think I will make a wiki/tutorial post @https://github.com/mrkn/pycall.rb since some people have same trouble as me

**Comment by buncis** [comment]
I manage to solved it by using approach mentioned in https://github.com/heroku/heroku-buildpack-python/issues/488#issuecomment-659579578

my `post_compile` scripts
```
set -e
buildpack_url=https://github.com/heroku/heroku-buildpack-python
buildpack_vsn=v197

# rebuild python if it's missing enable-shared
if  ! python3 -msysconfig | grep enable-shared \
    > /dev/null; then
  PYTHON_VERSION="$(< runtime.txt)"
  git clone -b "$buildpack_vsn" "$buildpack_url" _buildpack
  export WORKSPACE_DIR="$PWD/_buildpack/builds"
  rm -f .heroku/python/bin/python   # prevent failing ln after build
  sed -i 's!figure --pre!figure --enable-shared --pre!' \
    "$WORKSPACE_DIR"/runtimes/python3
  "$WORKSPACE_DIR/runtimes/$PYTHON_VERSION" /app/.heroku/python/
  rm -fr _buildpack
fi
```

latest build log and quick check

````
remote: Verifying deploy... done.
To https://git.heroku.com/xxxx.git
   7349abb..2a3c3cf  master -> master
~/xxxx (master) $ heroku run bash
 ›   Warning: heroku update available from 7.56.0 to 7.56.1.
Running bash on ⬢ xxxxx... up, run.8039 (Free)
~ $ python
Python 3.8.10 (default, Jul 30 2021, 07:53:46) 
[GCC 7.5.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sysconfig
>>> sysconfig.get_config_vars('Py_ENABLE_SHARED')
[1]
>>> exit()
```

**Comment by edmorley** [comment]
As of #1320 the Python binary build scripts now use `--enable-shared` for Python 3.10+.

Existing binaries are not being recompiled, so this will only take effect for future Python releases (ie: the future Python 3.10.5+) or for new stacks (ie: for Heroku-22's 3.10.*, once Heroku-22 is released).
